File: Detect Cycles in 2D Grid/Detect Cycles in 2D Grid.cpp
// Solution for Detect Cycles in 2D Grid in CPP

class Solution {
public:

    // Directions: left, down, up, right
    vector<vector<int>> directions = {{0, -1}, {1, 0}, {-1, 0}, {0, 1}};

    int rows, cols;

    // DFS to detect cycle
    bool dfs(vector<vector<bool>>& visited,
             vector<vector<char>>& grid,
             int currentRow,
             int currentCol,
             int parentRow,
             int parentCol) {

        // Mark current cell as visited
        visited[currentRow][currentCol] = true;

        // Explore all 4 directions
        for (const auto& dir : directions) {

            int nextRow = currentRow + dir[0];
            int nextCol = currentCol + dir[1];

            // Boundary check
            if (nextRow < 0 || nextRow >= rows || nextCol < 0 || nextCol >= cols)
                continue;

            // Only move to cells with same character
            if (grid[nextRow][nextCol] != grid[currentRow][currentCol])
                continue;

            // Skip the immediate parent cell (important to avoid false cycle)
            if (nextRow == parentRow && nextCol == parentCol)
                continue;

            // If already visited → cycle detected
            if (visited[nextRow][nextCol])
                return true;

            // Continue DFS
            if (dfs(visited, grid, nextRow, nextCol, currentRow, currentCol))
                return true;
        }

        return false;
    }

    bool containsCycle(vector<vector<char>>& grid) {

        rows = grid.size();
        cols = grid[0].size();

        vector<vector<bool>> visited(rows, vector<bool>(cols, false));

        // Start DFS from every unvisited cell
        for (int row = 0; row < rows; row++) {
            for (int col = 0; col < cols; col++) {

                if (!visited[row][col]) {
                    if (dfs(visited, grid, row, col, -1, -1)) {
                        return true;
                    }
                }
            }
        }

        return false;
    }
};

File: Detect Cycles in 2D Grid/Detect Cycles in 2D Grid.py
# Solution for Detect Cycles in 2D Grid in PY

class Solution:
    def containsCycle(self, grid):
        rows, cols = len(grid), len(grid[0])

        directions = [(0, -1), (1, 0), (-1, 0), (0, 1)]
        visited = [[False] * cols for _ in range(rows)]

        def dfs(r, c, pr, pc):
            visited[r][c] = True

            for dr, dc in directions:
                nr, nc = r + dr, c + dc

                # Boundary check
                if nr < 0 or nr >= rows or nc < 0 or nc >= cols:
                    continue

                # Same character only
                if grid[nr][nc] != grid[r][c]:
                    continue

                # Skip parent
                if nr == pr and nc == pc:
                    continue

                # Cycle detected
                if visited[nr][nc]:
                    return True

                if dfs(nr, nc, r, c):
                    return True

            return False

        for r in range(rows):
            for c in range(cols):
                if not visited[r][c]:
                    if dfs(r, c, -1, -1):
                        return True

        return False
